### ISSUE TITLE
steam.py instability

### ISSUE THREAD

**Issue by MichaReiser** [astral-sh/ty#3812]
### Summary

This reproduces the flaky `unused-type-ignore-comment` from
`steam/chat.py`. The underlying `invalid-argument-type` is stable with one
worker, but parallel checking sometimes resolves `Member.rank` as `Unknown`,
making the ignore appear unused.

```sh
for _ in {1..100}; do
    TY_MAX_PARALLELISM=4 /path/to/ty check steam --python-version 3.11 --output-format concise
done 2>&1 | grep -c unused-type-ignore-comment
```

[steam.zip](https://github.com/user-attachments/files/29134168/steam.zip)

**Comment by MichaReiser** [comment]
I post codex analysis here in case someone has an idea for how to fix this

> The issue is a nondeterministic cycle during class-base resolution:
> 
> ```python
> class Messageable(Protocol[MessageT]): ...
> 
> class WrapsUser(
>     PartialUser,
>     Messageable["UserMessage"],
> ): ...
> 
> class Member(WrapsUser, PartialMember):
>     def copy(self):
>         ProtoMember(self.rank)  # type: ignore
> ```
> 
> While resolving `Messageable["UserMessage"]`, ty can observe both a provisional value:
> 
> ```text
> Messageable[Unknown]
> ```
> 
> and a resolved value:
> 
> ```text
> Messageable[UserMessage]
> ```
> 
> Cycle recovery sometimes incorrectly combines these approximations into a semantic union. A union is not a valid class base, so MRO recovery inserts `Unknown`. Attribute lookup then returns `Unknown` for `self.rank` instead of finding `PartialMember.rank`.
> 
> That removes the argument-type error suppressed by the ignore, producing the flaky `unused-type-ignore-comment`.
> 
> Why does resolving `UserMessage` involve `GroupMember`?
> 
> Strictly speaking, `UserMessage` does not depend on `GroupMember`. The dependency comes from the granularity of ty’s module analysis:
> 
> ```python
> # message.py
> from .chat import Member
> 
> class GroupMember(Member): ...
> class UserMessage: ...
> ```
> 
> To resolve the forward reference `"UserMessage"`, ty must look up `UserMessage` in the `message` module. That lookup triggers module-level inference, which also processes `GroupMember`. Resolving `GroupMember`’s base requires `Member`, closing the cycle:
> 
> ```text
> Member
>   → Messageable["UserMessage"]
>   → analyze message.py
>   → GroupMember(Member)
>   → Member
> ```
> 
> So `GroupMember` is an implementation-level query dependency, not a semantic dependency of `UserMessage`. With perfectly fine-grained symbol resolution, looking up this empty `UserMessage` would not need to resolve `GroupMember`.

**Comment by MichaReiser** [comment]
I still see this on PRs, like here https://16fb055c.ty-ecosystem-ext.pages.dev/diff

**Comment by MichaReiser** [comment]
Here's a new reproducer 

Directory layout:

```text
known-specialization-union/
└── repro/
    ├── __init__.py
    ├── abc.py
    ├── chat.py
    ├── friend.py
    ├── message.py
    ├── user.py
    └── driver.py … driver7.py
```

`repro/__init__.py` is empty.

`repro/abc.py`

```python
from typing import TYPE_CHECKING, Generic, Protocol

from typing_extensions import TypeVar

if TYPE_CHECKING:
    from .message import Group


UserT = TypeVar("UserT", bound="PartialUser", default="PartialUser", covariant=True)
MessageT = TypeVar("MessageT", bound="Message", default="Message", covariant=True)


class PartialUser: ...


class Messageable(Protocol[MessageT]): ...


GroupT = TypeVar("GroupT", default="Group")


class Channel(Messageable[MessageT], Generic[MessageT, GroupT]): ...


ChannelT = TypeVar("ChannelT", bound=Channel, default=Channel, covariant=True)


class Message(Generic[UserT, ChannelT]): ...
```

`repro/chat.py`

```python
from typing import TYPE_CHECKING, Generic

from typing_extensions import TypeVar

from .abc import Channel, GroupT, Message, Messageable, PartialUser
from .user import User

if TYPE_CHECKING:
    from .message import GroupMessage, UserMessage


ChatT = TypeVar("ChatT", bound="Chat", default="Chat", covariant=True)
MemberT = TypeVar("MemberT")
AuthorT = TypeVar("AuthorT")


class WrapsUser(User if TYPE_CHECKING else PartialUser, Messageable["UserMessage"]): ...


class PartialMember(PartialUser):
    rank: int


class Member(WrapsUser): ...


class ChatMessage(
    Message[AuthorT, ChatT], Generic[AuthorT, MemberT, ChatT]
): ...  # ty: ignore[invalid-type-arguments]


ChatMessageT = TypeVar(
    "ChatMessageT",
    bound="GroupMessage",
    default="GroupMessage",
)


class Chat(
    Channel[ChatMessageT, GroupT]
): ...  # ty: ignore[invalid-type-arguments]


class ChatGroup(Generic[MemberT, ChatT]): ...
```

`repro/friend.py`

```python
from .chat import WrapsUser


class Friend(WrapsUser): ...
```

`repro/message.py`

```python
from typing import TYPE_CHECKING

from typing_extensions import TypeVar

from .abc import Message, PartialUser
from .chat import ChatGroup, ChatMessage, Member, PartialMember

if TYPE_CHECKING:
    from .friend import Friend
    from .user import ClientUser, User


class GroupMember(Member): ...


class Group(ChatGroup): ...


UserMessageAuthorT = TypeVar(
    "UserMessageAuthorT",
    bound=PartialUser,
    default="User | Friend | ClientUser",
    covariant=True,
)


class UserMessage(Message[UserMessageAuthorT]): ...


GroupMessageAuthorT = TypeVar("GroupMessageAuthorT", bound=PartialMember)


class GroupMessage(ChatMessage[GroupMessageAuthorT, GroupMember]): ...
```

`repro/user.py`

```python
from .abc import PartialUser


class User(PartialUser): ...


class ClientUser(PartialUser): ...
```

Each of `driver.py`, `driver1.py`, …, `driver7.py` has the same contents:

```python
from .chat import Member

x: type[Member] = Member
```

Run it from outside the Ruff checkout so Ruff’s root `pyproject.toml` is not discovered:

```sh
TY_MAX_PARALLELISM=4 /path/to/ty check repro \
    --python-version 3.11 \
    --output-format concise
```

The eight drivers are intentional scheduling inputs. On upstream `main`, repeated runs can still produce:

```text
repro/chat.py:17:57: warning[unsupported-base] Unsupported class base with type `<class 'Messageable[UserMessage[User | Friend | ClientUser]]'> | <class 'Messageable[UserMessage[User | ClientUser]]'>`
```

**Comment by MichaReiser** [comment]
Here a standalone test that doesn't require multithreading

```diff
diff --git a/crates/ty_python_semantic/src/types/tests.rs b/crates/ty_python_semantic/src/types/tests.rs
index 42e9a43356..53d381ad62 100644
--- a/crates/ty_python_semantic/src/types/tests.rs
+++ b/crates/ty_python_semantic/src/types/tests.rs
@@ -50,6 +50,14 @@ fn list_alias<'db>(db: &'db dyn Db, argument: Type<'db>) -> GenericAlias<'db> {
         .expect("a specialized `list` should be a generic alias")
 }
 
+fn iterable_alias<'db>(db: &'db dyn Db, argument: Type<'db>) -> GenericAlias<'db> {
+    KnownClass::Iterable
+        .to_specialized_class_type(db, &[argument])
+        .expect("`Iterable` should accept one type argument")
+        .into_generic_alias()
+        .expect("a specialized `Iterable` should be a generic alias")
+}
+
 fn oscillating_generic_alias_cycle_recover<'db>(
     db: &'db dyn Db,
     cycle: &salsa::Cycle,
@@ -76,6 +84,21 @@ fn oscillating_generic_alias(db: &dyn Db) -> Type<'_> {
     list_alias(db, argument).into()
 }
 
+#[salsa::tracked(
+    cycle_initial=|_, id| Type::divergent(id),
+    cycle_fn=oscillating_generic_alias_cycle_recover,
+)]
+fn oscillating_protocol_alias(db: &dyn Db) -> Type<'_> {
+    let previous = oscillating_protocol_alias(db);
+    let int: Type = iterable_alias(db, KnownClass::Int.to_instance(db)).into();
+
+    if previous == int {
+        iterable_alias(db, KnownClass::Str.to_instance(db)).into()
+    } else {
+        int
+    }
+}
+
 #[test]
 fn generic_alias_cycle_recovery_normalizes_same_origin_unknown_oscillation() {
     let db = setup_db();
@@ -101,6 +124,29 @@ fn generic_alias_cycle_recovery_rejects_unsafe_merges() {
     );
 }
 
+#[test]
+fn protocol_alias_cycle_recovery_produces_unsupported_class_base_type() {
+    let db = setup_db();
+    let int: Type = iterable_alias(&db, KnownClass::Int.to_instance(&db)).into();
+    let str: Type = iterable_alias(&db, KnownClass::Str.to_instance(&db)).into();
+
+    assert!(ClassBase::try_from_type(&db, int, None).is_some());
+    assert!(ClassBase::try_from_type(&db, str, None).is_some());
+
+    let recovered = oscillating_protocol_alias(&db);
+    let Type::Union(union) = recovered else {
+        panic!("cycle recovery should widen both protocol aliases");
+    };
+
+    assert_eq!(union.elements(&db).len(), 2);
+    assert!(union.elements(&db).contains(&int));
+    assert!(union.elements(&db).contains(&str));
+
+    // Each approximation is a valid base, but the recovered semantic union is not. This is the
+    // same failure shape as the schedule-dependent steam.py `unsupported-base` diagnostic.
+    assert!(ClassBase::try_from_type(&db, recovered, None).is_none());
+}
+
 /// All other tests also make sure that `Type::Todo` works as expected. This particular
 /// test makes sure that we handle `Todo` types correctly, even if they originate from
 /// different sources.
```

**Comment by charliermarsh** [comment]
Thanks. Are you working on fixing this?

**Comment by MichaReiser** [comment]
I have a diff with a fix but I can't claim to understand it (codex explanation makes sense to me).

**Comment by MichaReiser** [comment]
I abandon my attempt in https://github.com/astral-sh/ruff/pull/26633. The fix sounded plausible to me. But it also involves an area that I know too little about and familiarizing myself right now feels out of scope for a drive-by fix. So feel free to take over.
